### PR TITLE
Last Updated Timestamp

### PR DESCRIPTION
**Pull request by JulianForeman**
# Test Links:
[Percentile Calculator](https://wps-pr-1688.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-1688.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-1688.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-1688.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-1688.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=1108&f=c5&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN#state=2ec784ca-c46a-49d0-b2b3-1cf32a9015a2&session_state=7d9447c8-db66-4661-b4cb-03d2ac0d1d8f&code=32292df4-2bdf-4f90-a4a8-c8dbcda682a9.7d9447c8-db66-4661-b4cb-03d2ac0d1d8f.2b63f390-f3dc-43ae-89f2-016453863476)
[Fire Behaviour Advisory](https://wps-pr-1688.apps.silver.devops.gov.bc.ca/fire-behaviour-advisory)
[HFI Calculator](https://wps-pr-1688.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: web/src/features/hfiCalculator/components/testHelpers.ts [hunk 1]
```diff
@@ -37,6 +37,7 @@ export const buildStationDaily = (code: number, intensity_group = 1): StationDai
     intensity_group: intensity_group,
     sixty_minute_fire_size: 1,
     fire_type: 'fire',
-    date: DateTime.fromISO('2021-10-05T17:00:00.000-07:00')
+    date: DateTime.fromISO('2021-10-05T17:00:00.000-07:00'),
+    last_updated: new Date('2021-10-05T17:00:00.000-07:00')
```
Comment (Sybrand) [inline web/src/features/hfiCalculator/components/testHelpers.ts:41]: Can we be more consistent?
```suggestion
    last_updated: DateTime.fromISO('2021-10-05T17:00:00.000-07:00')
```

File: web/src/features/hfiCalculator/components/LastUpdatedHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        <span className={classes.container}>
+          <UpdateIcon className={classes.icon}></UpdateIcon>
+          <p className={classes.headerText}>Forecast last updated {dateString}</p>
+        </span>
+      </React.Fragment>
+    )
+  } else {
+    return (
+      <React.Fragment>
+        <p>Could not find last updated station</p>
```
Comment (Sybrand) [inline web/src/features/hfiCalculator/components/LastUpdatedHeader.tsx:61]: Is this technically correct? There's no update date - should this read something like "Could not find last updated date". Also I'm wondering - maybe a question for @tessmgood , but is there even any value in pushing this to the user? Maybe it should just log the error to the console, I just don't see a user being able to take any action on this.

File: web/src/features/hfiCalculator/pages/HfiCalculatorPage.tsx [hunk 3]
```diff
@@ -277,6 +278,7 @@ const HfiCalculatorPage: React.FunctionComponent = () => {
         </Container>
       ) : (
         <Container maxWidth={'xl'}>
+          <LastUpdatedHeader dailies={dailies} />
```
Comment (Sybrand) [inline web/src/features/hfiCalculator/pages/HfiCalculatorPage.tsx:281]: @tessmgood having the last update date here (as opposed to on the day in question) could cause some confusion. For example - if I click on any of the prep days, the date displayed here could possible relate to a different prep day.

**Review by Sybrand — COMMENTED** [review]
@JulianForeman super minor re. consistent use of date - and then mostly questions for @tessmgood

File: web/src/api/hfiCalculatorAPI.ts [hunk 4]
```diff
@@ -27,6 +27,7 @@ export interface StationDaily {
   sixty_minute_fire_size: number
   fire_type: string
   date: DateTime
+  last_updated: Date
```
Comment (conbrad) [inline web/src/api/hfiCalculatorAPI.ts:30]: Can we make this a `DateTime` like above?

File: web/src/features/hfiCalculator/components/LastUpdatedHeader.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    marginTop: '5px',
+    marginLeft: '5px'
+  },
+  headerText: {
+    marginLeft: '5px',
+    marginTop: '6px'
+  }
+})
+
+const findLastUpdate = (dailies: StationDaily[]) => {
```
Comment (conbrad) [inline web/src/features/hfiCalculator/components/LastUpdatedHeader.tsx:25]: This function can be simplified to:
```
maxBy(dailies.filter(daily => daily.status === 'FORECAST'), (forecastDaily) => forecastDaily.last_updated)
```

Or: 
```
const forecasts = dailies.filter(daily => daily.status === 'FORECAST')
maxBy(forecasts, (forecast) => forecast.last_updated)
```

With `lodash`

File: api/app/wildfire_one/schema_parsers.py [hunk 6]
```diff
@@ -273,7 +275,9 @@ def generate_station_daily(raw_daily,  # pylint: disable=too-many-locals
         sixty_minute_fire_size=sixty_minute_fire_size,
         fire_type=fire_type,
         error=raw_daily.get('observationValidInd', None),
-        error_message=raw_daily.get('observationValidComment', None)
+        error_message=raw_daily.get('observationValidComment', None),
+        last_updated=datetime.fromisoformat(raw_date_string[:len(
```
Comment (conbrad) [inline api/app/wildfire_one/schema_parsers.py:279]: What happens if we just set this as `raw_daily.get('updateDate', None)` ?

Comment (JulianForeman) [reply api/app/wildfire_one/schema_parsers.py:279]: It returns a date string that isn't iso compatible.

Comment (conbrad) [reply api/app/wildfire_one/schema_parsers.py:279]: Yuck. Ok, WFWX also returns an update timestamp in milliseconds, it's called `lastEntityUpdateTimestamp`.
https://chat.developer.gov.bc.ca/channel/wildfire-wfwx?msg=BvE29dpw6JuXTmd3i

So we can just do:
```
updateTimestamp = int(raw_daily.get('lastEntityUpdateTimestamp'))
last_updated=datetime.fromtimestamp(updateTimestamp / 1000, tz=timezone.utc)
```

**Review by tessmgood — APPROVED** [review]
Nice!

**Review by andrea-williams — CHANGES_REQUESTED** [review]
I randomly picked 2021/08/10 as date of interest - only Kamloops FC shows "Last updated" header - none of the other fire centres do.

Screen Shot 2022-02-03 at 3 10 36 PM

There's no error messages in the console to explain what's going on

**Review by andrea-williams — APPROVED** [review]
Sonarcloud is reporting a couple of code smells - please address those otherwise once this gets merged into main, all subsequent PRs will also have those code smells.
Otherwise, code all looks good!

File: web/cypress/integration/hfi-calculator-page.spec.ts [hunk 7]
```diff
@@ -6,7 +6,7 @@ function interceptDaily(fixturePath: string) {
     cy.intercept('GET', 'api/hfi-calc/daily*', req => {
       const date = new Date(Number(req.query['start_time_stamp']))
       dailies['dailies'].forEach(daily => {
-        daily['date'] = date.toISOString()
+        ;(daily['date'] = date.toISOString()), (daily['last_updated'] = date.toISOString())
```
Comment (conbrad) [inline web/cypress/integration/hfi-calculator-page.spec.ts:9]: Some invalid syntax here

**Review by conbrad — APPROVED** [review]
Just the syntax error in the cypress test, not sure why it's working though. Otherwise looks good to me.

**Review by Sybrand — APPROVED** [review]
Looks good - just address that syntax error @conbrad pointed out, and remove the console.log that you forgot in.

File: web/cypress/integration/hfi-calculator-page.spec.ts (hunk 7 above)
Comment (JulianForeman) [reply web/cypress/integration/hfi-calculator-page.spec.ts:9]: This is being place automatically when I save, I cannot remove it.

Comment (conbrad) [reply web/cypress/integration/hfi-calculator-page.spec.ts:9]: Try moving the stuff after the comma to a new line, VSCode breaks sometimes.

**Pull request merged**